spherical_splines/SHTOOLS/man/man1/makemaggrid2d.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MAKEMAGGRID2D 1"
.TH MAKEMAGGRID2D 1 "2009-08-18" "SHTOOLS 2.6" "SHTOOLS 2.6"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "MakeMagGrid2D"
.IX Header "MakeMagGrid2D"
.IP "MakeMagGrid2D \-" 17
.IX Item "MakeMagGrid2D -"
Create 2D cylindrical maps for all three vector components and total field from a set of magnetic potential spherical harmonic coefficients.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\s-1SUBROUTINE\s0 MakeMagGrid2D (" 27
.IX Item "SUBROUTINE MakeMagGrid2D ("
\&\s-1RAD\s0, \s-1PHI\s0, \s-1THETA\s0, \s-1TOTAL\s0, \s-1CILM\s0, R0, A, F, \s-1LMAX\s0, \s-1INTERVAL\s0, \s-1NLAT\s0, \s-1NLONG\s0, \s-1NORTH\s0, \s-1SOUTH\s0, \s-1EAST\s0, \s-1WEST\s0 )
.RS 4
.IP "REAL*8" 19
.IX Item "REAL*8"
\&\s-1RAD\s0(180/INTERVAL+1,\ 360/INTERVAL+1), \s-1PHI\s0(180/INTERVAL+1,\ 360/INTERVAL+1), \s-1THETA\s0(180/INTERVAL+1,\ 360/INTERVAL+1), \s-1TOTAL\s0(180/INTERVAL+1,\ 360/INTERVAL+1), \s-1CILM\s0(2,\ \s-1LMAX+1\s0,\ \s-1LMAX+1\s0), R0, A, F, \s-1INTERVAL\s0
.IP "\s-1INTEGER\s0" 19
.IX Item "INTEGER"
\&\s-1LMAX\s0, \s-1NLAT\s0, \s-1NLONG\s0
.IP "REAL*8, \s-1OPTIONAL\s0" 19
.IX Item "REAL*8, OPTIONAL"
\&\s-1NORTH\s0, \s-1SOUTH\s0, \s-1EAST\s0, \s-1WEST\s0
.RE
.RS 4
.RE
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fIMakeMagGrid2D\fR will create 2\-dimensional cylindrical maps of all three components and total field corresponding to a set of magnetic potential spherical harmonic coefficients. The magnetic potential is given by
.PP
V = R0 Sum_{l=1}^LMAX (R0/r)^{l+1} Sum_{m=\-l}^l C_{lm} Y_{lm}.
.PP
The output grids are in (geocentric) raster format possessing upper-left and lower-right coordinates of (90 N, 0 E) and (90 S, 360 E), respectively. If the optional parameters \s-1NORTH\s0, \s-1SOUTH\s0, \s-1EAST\s0 and \s-1WEST\s0 are specified, then the output grid will possess upper-left and lower-right coordinates of (\s-1NORTH\s0, \s-1WEST\s0) and (\s-1SOUTH\s0, \s-1EAST\s0), repsectively. The employed spherical harmonic normalization and Condon-Shortley phase convention is set to use Schmidt semi-normalized 4\-pi harmonics that exclude the phase of (\-1)^m. The field is calculated on the surface of an ellispoid with semis-major axis A and flattening F. Note that these calculations are strictly exact only when all points on the flattened ellipsoid are greater than the maximum radius of the planet.
.SH "ARGUMENTS"
.IX Header "ARGUMENTS"
.IP "\s-1RAD\s0" 11
.IX Item "RAD"
(output) REAL*8, \s-1DIMENSION\s0(180/INTERVAL+1, 360/INTERVAL+1)
.Sp
A 2D equally spaced map of the radial component of the magnetic field corresponding to the input spherical harmonic coefficients \s-1CILM\s0. The  array is in raster format with upper-left and lower-right coordinates of (90 N, 0 E) and (90 S, 360 E), respectively. Latitudinal coordinates are geocentric.
.IP "\s-1PHI\s0" 11
.IX Item "PHI"
(output) REAL*8, \s-1DIMENSION\s0(180/INTERVAL+1, 360/INTERVAL+1)
.Sp
A 2D equally spaced map of the phi (N\-S) component of the magnetic field corresponding to the input spherical harmonic coefficients \s-1CILM\s0.
.IP "\s-1THETA\s0" 11
.IX Item "THETA"
(output) REAL*8, \s-1DIMENSION\s0(180/INTERVAL+1, 360/INTERVAL+1)
.Sp
A 2D equally spaced map of the theta (E\-W) component of the magnetic field corresponding to the input spherical harmonic coefficients \s-1CILM\s0.
.IP "\s-1TOTAL\s0" 11
.IX Item "TOTAL"
(output) REAL*8, \s-1DIMENSION\s0(180/INTERVAL+1, 360/INTERVAL+1)
.Sp
A 2D equally spaced map of the total magnetic field corresponding to the input spherical harmonic coefficients \s-1CILM\s0.
.IP "\s-1CILM\s0" 11
.IX Item "CILM"
(input) REAL*8, \s-1DIMENSION\s0 (2, \s-1LMAX+1\s0, \s-1LMAX+1\s0)
.Sp
The real Schmidt semi-normalized spherical harmonic coefficients to be expanded in the space domain. The coefficients C1lm and C2lm refer to the \*(L"cosine\*(R" (Clm) and \*(L"sine\*(R" (Slm) coefficients, respectively, with Clm = \s-1CILM\s0(1, l+1, m+1) and Slm = \s-1CILM\s0(2, l+1, m+1). Alternatively, C1lm and C2lm correspond to the positive and negative order coefficients, respectively.
.IP "R0" 11
.IX Item "R0"
(input) REAL*8
.Sp
The reference radius of the spherical harmonic coefficients \s-1CILM\s0.
.IP "A" 11
.IX Item "A"
(input) REAL*8
.Sp
The semi-major axis of the ellipsoid used in evaluating the spherical harmonic coefficients in the space domain.
.IP "F" 11
.IX Item "F"
(input) REAL*8
.Sp
The flattening of the ellipsoid used in evaluating the spherical harmonic coefficients in the space domain.
.IP "\s-1LMAX\s0" 11
.IX Item "LMAX"
(input) \s-1INTEGER\s0
.Sp
The maximum spherical harmonic degree of the coefficients \s-1CILM\s0.
.IP "\s-1INTERVAL\s0" 11
.IX Item "INTERVAL"
(input) REAL*8
.Sp
The latitudinal and longitudinal spacing of the output grids \s-1RAD\s0, \s-1PHI\s0, \s-1THETA\s0 and \s-1TOTAL\s0.
.IP "\s-1NLAT\s0" 11
.IX Item "NLAT"
(output) \s-1INTEGER\s0
.Sp
The number of latitudinal samples. Both 90 N and 90 S are included.
.IP "\s-1NLONG\s0" 11
.IX Item "NLONG"
(output) \s-1INTEGER\s0
.Sp
The number of longitudinal samples. Both 0 and 360 E are included.
.IP "\s-1NORTH\s0" 11
.IX Item "NORTH"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The maximum latitude of the output raster grid, in degrees. The default is 90 degrees.
.IP "\s-1SOUTH\s0" 11
.IX Item "SOUTH"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The minimum latitude of the output raster grid, in degrees. The default is \-90 degrees.
.IP "\s-1EAST\s0" 11
.IX Item "EAST"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The maximum longitude of the output raster grid, in degrees. The default is 360 degrees.
.IP "\s-1WEST\s0" 11
.IX Item "WEST"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The minimum longitude of the output raster grid, in degrees. The default is 0 degrees.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
<http://www.ipgp.jussieu.fr/~wieczor/SHTOOLS/SHTOOLS.html>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2005\-2007 by Mark Wieczorek <wieczor@ipgp.fr>.
.PP
This is free software; you can distribute and modify it under the terms of the revised \s-1BSD\s0 license.


spherical_splines/SHTOOLS/src/MakeMagGrid2D.f95
subroutine MakeMagGrid2D(rad, phi, theta, total, cilm, r0, a, f, lmax, interval, nlat, nlong, &
	north, south, east, west)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!	Given the Schmidt normalized Spherical Harmonic coefficients cilm, this subroutine
!	will compute a 2D grid with equal latitude and longitude spacings for the radial, 
!	phi and theta directions, as well as the total field.  The field is calculated 
!	on a flattended ellipsoid with semi-major axis A and flattening F.
!	
!	Note that since this is NOT done using FFTs, this routine is relatively SLOW!
!
!	The value at grid(1,1) correspons to 90 degrees latitude and 0 degress
!	longitude. Longitude points are calculated from 0 to 360, with both points 
!	being included. If the optional parameters NORTH, SOUTH, EAST and WEST are specified, 
!	the upper-left and lower right coordinates of the output grid are (NORTH, WEST) and 
!	(SOUTH, EAST), respectively.
!
!	Calling Parameters:
!		IN
!			cilm:		Schmidt Normalized spherical harmonic coefficients with 
!					dimensions (2, lmax+1, lmax+1).
!			r0:		Reference radius of cilm coeficients.
!			a:		Semi-major axis of the reference ellipsoid.
!			f:		Flattening of reference radius: (R_eq - R_p)/R_eq, where R_eq = A.
!			lmax:		Maximum spherical harmonic degree of expansions to be performed.
!			interval:	Spacing of output grid in DEGREES.
!
!		OUT
!			rad:	 	Grid of radial magnetic field.
!			phi:		Grid of magnetic field in phi direction.
!			theta:		Grid of magnetic field in theta direction.
!			total:		Grid of magnetic intensity.
!			nlat:		Number of latitude points for the grid.
!			nlong:		Number of longitude points for the grid.
!		OPTIONAL (IN)
!			north		Maximum latitude to compute, in degrees.
!			south		Minimum latitude to compute, in degrees.
!			east		Maximum longitude to compute, in degrees.
!			west		Minimum latitude to compute, in degrees.
!
!	Notes:
!		1.	If lmax is greater than the the maximum spherical harmonic
!			degree of the input file, then this file will be ZERO PADDED!
!			(i.e., those degrees after lmax are assumed to be zero).
!		2. 	Latitude is geocentric latitude.
!
!		Dependencies:	PlmSchmidt_d1, PlSchmidt, PlmIndex
!
!	Written by Mark Wieczorek (2003)
!
!	August 2005 - 	sines and cosines are now precomputed, leading to a savings in time
!			by a factor of up to 15 (if the code is compiled with normal
!			optimizations.
!	May 31, 2006 -  Modified so that the Condon-Shortley phase is never used, regardless of 
!			what the default value of CSPHASE_DEFAULT is.
!	June 21, 2007 - Modified routine such that the field is calculated on a flattened
!			ellisoid, using exact equations for the ellipsoid. Previously, only
!			the degree-2 Legendre expansion was utilized. It is now necessary to
!		 	input the semi-major axis of the ellispoid instead of the mean ellipsoid
!			radius.
!	August 19, 2007	Added option to calculate subdomains of a raster grid.
!
!	Copyright (c) 2005-2007, Mark A. Wieczorek
!	All rights reserved.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	use SHTOOLS, only: PlmSchmidt_d1, PlSchmidt, PlmIndex

	implicit none
	
	real*8, intent(in) :: 	cilm(:,:,:), interval, r0, a, f
	real*8, intent(out) :: 	rad(:,:), phi(:,:), theta(:,:), total(:,:)
	integer, intent(in) :: 	lmax
	integer, intent(out) :: nlat, nlong
	real*8, intent(in), optional :: north, south, east, west
	integer :: 		l, m, j, k, index, l1, m1, lmax_comp, temp, astat(4)
	real*8 :: 		pi, latmax, latmin, longmin, longmax, lat, &
				longitude, x, intervalrad, r_ex, coef
	real*8, allocatable ::	pl(:), dpl(:), cosm(:,:), sinm(:,:)
		
    	
    	temp = 0
	if (present(north)) temp = temp + 1
	if (present(south)) temp = temp + 1
	if (present(east)) temp = temp + 1
	if (present(west)) temp = temp + 1
	
	if (temp /=0 .and. temp /=4) then
		print*, "Error --- MakeMagGrid2d"
		print*, "The optional parameters NORTH, SOUTH, EAST, and WEST must all be specified", &
			present(north), present(south), present(east), present(west)
		stop
	endif
	
	if (temp == 4) then
		latmax = north
		latmin = south
		longmin = west
		longmax = east
		
		if (latmax < latmin) then
			print*, "Error --- MakeMagGrid2d"
			print*, "NORTH must be larger than SOUTH."
			print*, "NORTH = ", latmax
			print*, "SOUTH = ", latmin
			stop
		endif
		
		if (latmax > 90.0d0 .or. latmin < -90.0d0) then
			print*, "Error --- MakeMagGrid2d"
			print*, "NORTH and SOUTH must lie between 90 and -90."
			print*, "NORTH = ", latmax
			print*, "SOUTH = ", latmin
			stop
		endif
		
		if (longmin > longmax) longmax = longmax + 360.0d0
		
	else
		latmax = 90.0d0
		latmin = -90.0d0
		longmin = 0.0d0
		longmax = 360.d0
	endif
	
	nlat = (latmax - latmin) / interval + 1
	nlong = (longmax - longmin) / interval + 1
    	
    	
 	if (size(rad(:,1)) < nlat .or. size(rad(1,:)) < nlong ) then
		print*, "Error --- MakeMagGrid2D"
		print*, "RAD must be dimensioned ( (LATMAX-LATMIN)/INTERVAL+1, (LONGMAX-LONGMIN)/INTERVAL+1 ) where"
		print*, "INTERVAL = ", interval
		print*, "LATMAX = ", latmax
		print*, "LATMIN = ", latmin
		print*, "LONGMIN = ", longmin
		print*, "LONGMAX = ", longmax
		print*, "Input array is dimensioned ", size(rad(:,1)), size(rad(1,:))
		stop
 	elseif (size(phi(:,1)) < nlat .or. size(phi(1,:)) < nlong ) then
		print*, "Error --- MakeMagGrid2D"
		print*, "PHI must be dimensioned ( (LATMAX-LATMIN)/INTERVAL+1, (LONGMAX-LONGMIN)/INTERVAL+1 ) where"
		print*, "INTERVAL = ", interval
		print*, "LATMAX = ", latmax
		print*, "LATMIN = ", latmin
		print*, "LONGMIN = ", longmin
		print*, "LONGMAX = ", longmax
		print*, "Input array is dimensioned ", size(phi(:,1)), size(phi(1,:))
		stop
 	elseif (size(theta(:,1)) < nlat .or. size(theta(1,:)) < nlong ) then
		print*, "Error --- MakeMagGrid2D"
		print*, "THETA must be dimensioned ( (LATMAX-LATMIN)/INTERVAL+1, (LONGMAX-LONGMIN)/INTERVAL+1 ) where"
		print*, "INTERVAL = ", interval
		print*, "LATMAX = ", latmax
		print*, "LATMIN = ", latmin
		print*, "LONGMIN = ", longmin
		print*, "LONGMAX = ", longmax
		print*, "Input array is dimensioned ", size(theta(:,1)), size(theta(1,:))
		stop
 	elseif (size(total(:,1)) < nlat .or. size(total(1,:)) < nlong ) then
		print*, "Error --- MakeMagGrid2D"
		print*, "TOTAL must be dimensioned ( (LATMAX-LATMIN)/INTERVAL+1, (LONGMAX-LONGMIN)/INTERVAL+1 ) where"
		print*, "INTERVAL = ", interval
		print*, "LATMAX = ", latmax
		print*, "LATMIN = ", latmin
		print*, "LONGMIN = ", longmin
		print*, "LONGMAX = ", longmax
		print*, "Input array is dimensioned ", size(total(:,1)), size(total(1,:))
		stop
	elseif (size(cilm(:,1,1)) < 2 .or. size(cilm(1,:,1)) < lmax+1 .or. size(cilm(1,1,:)) < lmax+1) then
		print*, "Error --- MakeMagGrid2D"
		print*, "CILM must be dimensioned as (2, LMAX+1, LMAX+1) where LMAX is ", lmax
		print*, "Input dimension is ", size(cilm(:,1,1)), size(cilm(1,:,1)), size(cilm(1,1,:))
		stop
	endif
	
	allocate(pl((lmax+1)*(lmax+2)/2), stat = astat(1))
	allocate(cosm(int(360./interval + 1), lmax+1), stat = astat(2))
	allocate(sinm(int(360./interval + 1), lmax+1), stat = astat(3))
	allocate(dpl((lmax+1)*(lmax+2)/2), stat = astat(4))
	if (astat(1) /= 0 .or. astat(2) /= 0 .or. astat(3) /=0 .or. astat(4) /= 0) then
		print*, "Error --- MakeMagGrid2D"
		print*, "Problem allocating arrays PL, COSM, SINM, and DPL", astat(1), astat(2), astat(3), astat(4)
		stop
	endif
	
	pi = acos(-1.0d0)
	rad = 0.0d0
	phi = 0.0d0
	theta = 0.0d0
	total = 0.0d0

	intervalrad = interval*pi/180.0d0
	
	lmax_comp = min(lmax, size(cilm(1,1,:))-1)
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Loop over latitude. Precompute sin and cosine terms.
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	do k=1, nlong
	
		longitude = longmin*pi/180.0d0 + dble(k-1)*intervalrad
		do m=0, lmax
			cosm(k, m+1) = cos(m*longitude)
			sinm(k, m+1) = sin(m*longitude)
		enddo
		
	enddo

	do j=1, nlat
	
		lat = latmax - dble(j-1)*interval
		x = sin(lat*pi/180.0d0)
		
		if (lat == 90.0d0 .or. lat == -90.0d0) then
		
			call PlSchmidt(pl, lmax_comp, x)
	
			r_ex = a * (1.0d0 - f)	! Reference ellipsoid to calculate field			

			coef = (r0/r_ex)**2
			
			do l = 1, lmax_comp, 1
				l1 = l+1
				coef = coef * (r0/r_ex)
				rad(j,1) = rad(j,1) + dble(l+1) * cilm(1,l1,1) * pl(l1) * coef
			enddo
	
			phi(j,1:nlong) = 0.0d0 
			rad(j,2:nlong) = rad(j,1)
			theta(j,1:nlong) = 0.0d0
			
		else
		
			r_ex = a**2 * (1.0d0 + tan(lat*pi/180.0d0)**2) / &
				(1.0d0  + tan(lat*pi/180.0d0)**2 / (1.0d0 - f)**2 )
			r_ex = sqrt(r_ex)

			call PlmSchmidt_d1(pl, dpl, lmax_comp, x, csphase = 1)
			dpl = - dpl * cos(lat*pi/180.0d0)	! convert derivative with respect to z to
								! a derivative with respect to theta.
			do k = 1, nlong
					
				coef = (r0/r_ex)**2
					
				do l = 1, lmax_comp, 1
					l1 = l+1
					coef = coef * (r0/r_ex)
					do m = 0, l
						index = PlmIndex(l,m)
						m1 = m+1
						rad(j,k) = rad(j,k) + dble(l+1) * (cilm(1,l1,m1)*cosm(k,m1) + &
							cilm(2,l1,m1)*sinm(k, m1) ) * pl(index) * coef
						phi(j,k) = phi(j,k) + dble(m)*(cilm(1,l1,m1)*sinm(k,m+1) - &
							cilm(2,l1,m1)*cosm(k, m+1) ) * pl(index) * coef &
							/ cos(lat*pi/180.0d0)
						theta(j,k) = theta(j,k) - (cilm(1,l1,m1)*cosm(k,m1) + &
							cilm(2,l1,m1)*sinm(k,m1) ) * dpl(index)  &
							* coef
						
					enddo
				enddo
			
			enddo
			
		endif
		
	enddo
	

	total(1:nlat, 1:nlong) = sqrt( rad(1:nlat,1:nlong)**2 + phi(1:nlat,1:nlong)**2 + theta(1:nlat,1:nlong)**2)
	
	! deallocate memory
	call PlmSchmidt_d1(pl, dpl, -1, x, csphase = 1)
	deallocate(pl)
	deallocate(cosm)
	deallocate(sinm)
	deallocate(dpl)

end subroutine MakeMagGrid2D

